### ISSUE TITLE
Generating URLs for a different action on the same controller reuses ambient values

### ISSUE THREAD

**Issue by danports** [dotnet/aspnetcore#18400]
The [documentation](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/routing?view=aspnetcore-3.1#endpoint-routing-differences-from-earlier-versions-of-routing) states that "In ASP.NET Core 2.2 or later, both [conventional and attribute] routing invalidate [ambient] values when linking to another action." However, this only seems to be the case for actions on another controller. I am unsure whether this is a bug or the expected behavior with unclear or incorrect documentation.

Repro: Run [this sample project](https://github.com/danports/aspnetcore-routingtest), click the Orders tab in the navigation bar, select an order from the list, and then observe that the URL for the Orders tab in the navigation bar _as generated on the order details page_ is `/Orders/5/Index` rather than the expected `/Orders` (since the documentation says the ambient value for `id` shouldn't be reused since this is a different action). Note also that the Home and Privacy navigation links on the order details page do not reuse ambient values and do not include `id` in their URLs.

**Comment by javiercn** [comment]
@danports thanks for contacting us.

```
endpoints.MapControllerRoute("itemRoute", "{controller}/{id:int}/{action=Details}");
endpoints.MapControllerRoute("default", "{controller=Home}/{action=Index}/{id?}");
```

```
<li><a asp-action="Details" asp-route-id="@i">@i</a></li>
```


I think this has to do with the fact that you are specifying the details. If you remove the `asp-action="Details"` it should give you the right outcome. Is that not the case?

**Comment by danports** [comment]
Removing `asp-action="Details"` breaks the links from the order list page to the details page - instead of generating `/Orders/4`, it generates `/Orders/4/Index`. And if you do navigate directly to the details page, the link back to the orders list in the navigation bar is still wrong.

**Comment by mkArtakMSFT** [comment]
@rynowak can you please look into this? Thanks!

**Comment by danports** [comment]
Any updates on this? Would be helpful to know if what I'm seeing is the expected/desired behavior or a bug.

**Comment by danports** [comment]
Alright, I dug through the code until I found [the bits that do ambient value invalidation](https://github.com/dotnet/aspnetcore/blob/1c126ab773059d6a5899fc29547cb86ed49c46bf/src/Http/Routing/src/Template/TemplateBinder.cs#L163) and I think the code is behaving as designed, though the documentation for ambient value invalidation is a bit confusing. I think the intent of the algorithm is to keep things DRY by avoiding the need to specify every route value for every link, which means that cases like this where you don't want ambient values to be inherited require a bit more work. Including `asp-route-id="@null"` in the Orders link in the layout page in the sample I created earlier does cause the correct URL to be generated on the order details page. So I can deal with this by adding that attribute to any link where the target action does not expect an `id` parameter.

To sum up then, a couple of suggestions:
1. Improve the documentation for ambient value invalidation, e.g. by explaining how the left to right matching works and showing a few more nontrivial examples, including some with non-default route patterns and one where an ambient value is overridden with `null`, which was the workaround I needed but didn't see until I read the framework code.
2. Make endpoint routing smarter: Not sure how feasible this is, but it would be nice if the framework knew that the Index action didn't take an `id` parameter and preferred to generate links for that action using routes that didn't have an `id` parameter.
